### ISSUE TITLE
Dependency insight report doesn't support project dependency queries

### ISSUE THREAD

**Issue by melix** [gradle/gradle#6857]
### Context

When specifying the `--dependency` option, it should be possible to use a project name, for project dependencies. However, we don't match on project dependencies. There are at least 2 places where this happens:

https://github.com/melix/gradle/blob/a3c5a35df0b3c44dc990eb6a2b53a92977fcedb0/subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/dsl/DependencyResultSpec.java#L47-L51

Here we need to add something like:

```
 else if (requested instanceof ProjectComponentSelector) {
            ProjectComponentSelector project = (ProjectComponentSelector)requested;
            String requestedCandidate = project.getProjectPath();
            return requestedCandidate.contains(stringNotation);
        }
```

And https://github.com/melix/gradle/blob/a3c5a35df0b3c44dc990eb6a2b53a92977fcedb0/subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/graph/nodes/UnresolvedDependencyEdge.java#L38-L40

where the cast actually becomes unsafe. By baking in a fake `ModuleComponentIdentifier`, we can get through:

```
 else if (attempted instanceof ProjectComponentSelector){
            actual = DefaultModuleComponentIdentifier.newId(DefaultModuleIdentifier.newId("", ((ProjectComponentSelector) attempted).getProjectPath()), "1.0");
        } else {
            actual = null;
        }
```

Then it's possible to show the result: https://e.grdev.net/s/zddtq53t7q3uk/console-log#L360-L378

However it would be better if `UnresolvedDependencyEdge` didn't assume a `ModuleComponentIdentifier`.

This is more problematic now that project dependencies can be resolved via variant aware dependency management (attribute based).

### Your Environment

 * Build scan URL:

**Comment by big-guy** [comment]
@bigdaz this doesn't seem to quite work.
When there's a failed _project_ dependency, dependencyInsight explodes here:
https://github.com/gradle/gradle/blob/a3c5a35df0b3c44dc990eb6a2b53a92977fcedb0/subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/graph/nodes/UnresolvedDependencyEdge.java#L38

You can reproduce with the gradle/gradle build: `local_gradle distributions:dependencyInsight --configuration gradleDocumentation --dependency :docs`

**Comment by bigdaz** [comment]
Thanks @big-guy. I seem to be working on too many PRs at once, and the fix for this is in another (yet-to-be-merged) one: https://github.com/gradle/gradle/pull/7629/files#diff-724f362d492683678ef372db006e378aR40

I'll add an integration test for this case.

**Comment by bigdaz** [comment]
@big-guy this issue has been fixed by e30cf982e7262ab21246459eaef98fe6b8325e01. The output of  `gradle distributions:dependencyInsight --configuration gradleDocumentation --dependency :docs` is now:
```
> Task :distributions:dependencyInsight
project :docs FAILED
   Failures:
      - Could not resolve project :docs.:
          - Cannot choose between the following variants of project :docs:
              - gradleFullDocsElements
              - gradleGettingStartedElements
              - gradleSamplesElements
            All of them match the consumer attributes:
              - Variant 'gradleFullDocsElements':
                  - Required minified 'true' and found compatible value 'true'.
                  - Required org.gradle.usage 'docs' and found compatible value 'docs'.
                  - Found type 'full-docs' but wasn't required.
              - Variant 'gradleGettingStartedElements':
                  - Required minified 'true' and found compatible value 'true'.
                  - Required org.gradle.usage 'docs' and found compatible value 'docs'.
                  - Found type 'getting-started' but wasn't required.
              - Variant 'gradleSamplesElements':
                  - Required minified 'true' and found compatible value 'true'.
                  - Required org.gradle.usage 'docs' and found compatible value 'docs'.
                  - Found type 'samples' but wasn't required.

project :docs FAILED
\--- gradleDocumentation
```

Thanks for reporting the issue. I think this makes `dependencyInsight` much more useful for diagnosing attribute matching issues with project dependencies.

**Comment by big-guy** [comment]
Thanks @bigdaz, this is great
